build/toml-patch/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+                    match u {
+                        // Single values replace the previous value
+                        Value::Float(..)
+                        | Value::String(..)
+                        | Value::Integer(..)
+                        | Value::Boolean(..)
+                        | Value::Datetime(..) => *u = v.clone(),
+
+                        // Inline tables are not yet supported, but should be
+                        // merged once we get around to implementing it
+                        Value::InlineTable(..) => {
+                            bail!(
+                                "patching inline tables is not yet implemented"
+                            );
+                        }
+                        // Arrays are extended
+                        Value::Array(u) => {
+                            u.extend(v.as_array().unwrap().iter().cloned());
+                        }
+                    }
+                }
+                Item::Table(u) => {
+                    // Recurse!
+                    merge_toml_tables(u, v.as_table_mut().unwrap())?;
+                }
+                Item::ArrayOfTables(arr) => {
+                    // Compute an offset based on table position
+                    let mut visitor = TableRangeVisitor::default();
+                    visitor.visit_array_of_tables(arr);
+                    let range_in = visitor.range.unwrap();
+                    let last = range_in.end;
+
+                    let mut visitor = TableRangeVisitor::default();
+                    visitor.visit_item(v);
+                    let start =
+                        visitor.range.map(|r| r.start as isize).unwrap();
+                    let offset = last as isize - start as isize;
+
+                    // Apply that offset to the incoming tables
+                    let mut visitor = TableShiftVisitor { offset };
+                    visitor.visit_item_mut(v);
+
+                    // Merge by extending the table array
+                    arr.extend(v.as_array_of_tables().unwrap().iter().cloned());
+                }
+            }
+        } else {
+            let mut visitor = TableRangeVisitor::default();
+            visitor.visit_table(original);
+            let last = visitor.range.unwrap().end;
+
+            let mut visitor = TableRangeVisitor::default();
+            visitor.visit_item(v);
+            let start = visitor.range.map(|r| r.start as isize).unwrap_or(0);
+            let offset = last as isize - start as isize;
+
+            // Apply that offset to the incoming tables
+            let mut visitor = TableShiftVisitor { offset };
+            visitor.visit_item_mut(v);
+
+            // Merge by inserting the new element
+            original.insert(k.get(), v.clone());
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use indoc::indoc;
+
+    fn patch_and_compare(a: &str, b: &str, out: &str) {
```

Comment:
love this style of testing, guessing you read that matklad blog post too haha

Reply:
I haven't, actually!  I see a few, which ones are recommended reading?

Reply:
This post specifically advocates doing exactly this https://matklad.github.io/2021/05/31/how-to-test.html